### PR TITLE
fix(dsstore): persist inherited LTS trie

### PR DESCRIPTION
**Pull request by keynslug**
Release version: v/e5.7.1

## Summary

Before this commit, inherited trie was actually only kept in memory cache.

Also simplify storage backend behaviour around inheriting previous generation's legacy.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] ~~Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~~
- [x] Schema changes are backward compatible

### CODE REVIEW

File: apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl [hunk 1]
```diff
@@ -23,6 +23,8 @@
 -include_lib("snabbkaffe/include/snabbkaffe.hrl").
 -include_lib("stdlib/include/assert.hrl").
 
+-import(emqx_utils_conv, [bin/1]).
```
Comment (ieQu1) [inline apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:26]: Better to avoid imports.

Comment (keynslug) [reply apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl]: Dropped in 10e78ea194e4b8f5592b198c9fff79723905985e.

File: apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl [hunk 2]
```diff
@@ -177,20 +177,33 @@ t_new_generation_inherit_trie(_Config) ->
...
-            Batch = [
-                begin
-                    Topic = emqx_topic:join(["wildcard", integer_to_binary(I), "suffix", Suffix]),
-                    {TS, make_message(TS, Topic, integer_to_binary(TS))}
-                end
+            Batch1 = [
+                {TS, make_message(TS, make_topic(wildcard, I, suffix, Suffix), bin(TS))}
              || I <- lists:seq(1, 200),
-                TS <- Timestamps,
+                TS <- lists:seq(0, 1_000, 10),
```
Comment (ieQu1) [inline apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:183]: It's pretty confusing that `TS` is used for both timestamp and as a topic level. I think it's better to use a constant timestamp for all messages (I believe it's irrelevant in this test), and rename `TS` to `J`.

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl [hunk 3]
```diff
@@ -194,25 +193,19 @@
...
 %% Create rocksdb column families.
--callback create(shard_id(), rocksdb:db_handle(), gen_id(), Options :: map()) ->
+-callback create(
+    shard_id(),
+    rocksdb:db_handle(),
+    gen_id(),
+    Options :: map(),
+    PrevRuntimeData :: term()
+) ->
     {_Schema, cf_refs()}.
```
Comment (ieQu1) [inline apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:209]:
```suggestion
    Schema | undefined
) ->
    {Schema, cf_refs()}.
```

File: apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl [hunk 4]
```diff
@@ -475,6 +485,12 @@ make_message(PublishedAt, Topic, Payload) when is_binary(Topic) ->
         payload = Payload
     }.
 
+make_topic(T1, T2, T3, T4) ->
+    make_topic([T1, T2, T3, T4]).
```
Comment (ieQu1) [inline apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:489]: This function looks very specific to a particular testcase. Please remove it and use `make_topic/1` with list argument.

File: apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:183]: It's actually not used as a topic level, but it indeed should be irrelevant.

File: apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl (hunk 3 above)
Comment (keynslug) [reply apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl:209]: The last argument is actually `RuntimeData` (i.e. the result of `open/5`), not `Schema`.

**Review by ieQu1 — APPROVED** [review]

File: apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl [hunk 5]
```diff
@@ -211,10 +224,7 @@ t_replay(_Config) ->
     ok = emqx_ds_storage_layer:store_batch(?SHARD, Batch1, []),
     %% Create wildcard topics `wildcard/+/suffix/foo' and `wildcard/+/suffix/bar':
     Batch2 = [
-        begin
-            Topic = emqx_topic:join(["wildcard", integer_to_list(I), "suffix", Suffix]),
-            {TS, make_message(TS, Topic, integer_to_binary(TS))}
-        end
+        {TS, make_message(TS, make_topic([wildcard, I, suffix, Suffix]), bin(TS))}
```
Comment (ieQu1) [inline apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:227]: Same here, TS is probably not needed.

Comment (keynslug) [reply apps/emqx_durable_storage/test/emqx_ds_storage_bitfield_lts_SUITE.erl:227]: Had a different impression actually: sounds important for a replay test to have messages spread both in time and in "topic space" (here, `TS` is not part of a topic).

Anyway, it'd probably be better to tackle the logic of unrelated testcase in a separate PR.

**Review by ieQu1 — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-2645)

**Pull request merged**